In [2]:
from sqlalchemy import create_engine

engine = create_engine(
    "postgresql+psycopg2://admin:admin@localhost:5432/bankdb"
)

print("Ligação criada com sucesso")

Ligação criada com sucesso


In [4]:
import pandas as pd

df = pd.read_sql(
    "SELECT * FROM bank_marketing",
    engine
)

print("Dimensão do dataset:")
print(df.shape)

Dimensão do dataset:
(45214, 18)


# 1. Dimensão do Dataset

## Observação

O dataset contém 45.214 registos e 18 variáveis após o processo de limpeza e carregamento para PostgreSQL.

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45214 entries, 0 to 45213
Data columns (total 18 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   unnamed: 0  45214 non-null  int64  
 1   age         45214 non-null  int64  
 2   job         45214 non-null  object 
 3   marital     45214 non-null  object 
 4   education   45214 non-null  object 
 5   default     45214 non-null  object 
 6   balance     45214 non-null  float64
 7   housing     45214 non-null  object 
 8   loan        45214 non-null  object 
 9   contact     45214 non-null  object 
 10  day         45214 non-null  int64  
 11  month       45214 non-null  object 
 12  duration    45214 non-null  float64
 13  campaign    45214 non-null  float64
 14  pdays       45214 non-null  float64
 15  previous    45214 non-null  int64  
 16  poutcome    45214 non-null  object 
 17  y           45214 non-null  object 
dtypes: float64(4), int64(4), object(10)
memory usage: 6.2+ MB


## 2. Estrutura do Dataset

O dataset contém 18 colunas e 45.214 registos.

Foram identificados:

- 10 variáveis categóricas (object)
- 4 variáveis numéricas inteiras (int64)
- 4 variáveis numéricas decimais (float64)

Não foram encontrados valores nulos após o processo de limpeza dos dados.

Observa-se ainda a existência da coluna `unnamed: 0`, que aparenta corresponder a um índice importado do ficheiro CSV original.

In [6]:
df.describe()

,unnamed: 0,age,balance,day,duration,campaign,pdays,previous
count,45214.000000,45214.000000,45214.000000,45214.000000,45214.000000,45214.000000,45214.000000,45214.000000
mean,22606.500000,40.891096,1338.604392,15.806299,257.718981,2.758194,39.815057,0.580285
std,13052.301872,10.537422,2993.997526,8.323983,257.019318,3.087892,99.658218,2.303369
min,0.000000,-5.000000,-8019.000000,0.000000,-30.000000,0.000000,-1.000000,0.000000
25%,11303.250000,33.000000,80.000000,8.000000,104.000000,1.000000,-1.000000,0.000000
50%,22606.500000,39.000000,450.000000,16.000000,180.000000,2.000000,-1.000000,0.000000
75%,33909.750000,48.000000,1385.000000,21.000000,317.000000,3.000000,-1.000000,0.000000
max,45213.000000,200.000000,102127.000000,45.000000,4918.000000,63.000000,854.000000,275.000000


## 3. Estatísticas Descritivas

A análise estatística das variáveis numéricas permite identificar possíveis anomalias e compreender melhor a distribuição dos dados.

Observações:

- A idade média dos clientes é aproximadamente 40,9 anos.
- O saldo médio das contas é cerca de 1338 unidades monetárias.
- A duração média dos contactos é aproximadamente 258 segundos.
- Existem valores extremos em várias variáveis, nomeadamente na idade, saldo e duração dos contactos.
- Foram identificados valores negativos em algumas variáveis, o que poderá indicar erros ou inconsistências presentes no dataset original.

Foi possível identificar valores potencialmente inválidos, como idades negativas (-5 anos), idades excessivamente elevadas (200 anos) e durações negativas de contacto (-30 segundos). Estes registos evidenciam a presença de dados anómalos no dataset original.



In [7]:
df["y"].value_counts()

y
no                39139
yes                5112
not-subscribed      651
?                   226
subscribed           86
Name: count, dtype: int64

## 4. Distribuição da Variável Alvo (y)

A variável `y` representa o resultado da campanha de marketing bancário.

Foram identificados valores esperados (`yes` e `no`) mas também valores inconsistentes (`subscribed`, `not-subscribed` e `?`).

Esta situação sugere que poderão existir oportunidades adicionais de normalização dos dados após a fase de ingestão e validação.

In [8]:
df["job"].value_counts().head(10)

job
blue-collar      10517
management        9103
technician        7311
admin.            5064
services          4069
retired           2225
self-employed     1546
entrepreneur      1457
unemployed        1278
housemaid         1214
Name: count, dtype: int64

## 5. Distribuição das Profissões

A análise da variável `job` demonstra que a maioria dos clientes pertence às profissões:

- Blue-collar (10.517 clientes)
- Management (9.103 clientes)
- Technician (7.311 clientes)

Estas três categorias representam uma parte significativa da população presente no dataset.

As profissões menos representadas incluem housemaid, unemployed e entrepreneur.

In [9]:
df["marital"].value_counts()

marital
married     27336
single      12543
divorced     5109
?             226
Name: count, dtype: int64

## 6. Estado Civil dos Clientes

A maioria dos clientes é casada (married), seguida por clientes solteiros (single) e divorciados (divorced).

Foi também identificado um conjunto de 226 registos com o valor "?", indicando informação desconhecida ou não tratada durante o processo de limpeza.

Este resultado reforça a necessidade de validação adicional dos dados após a ingestão na base de dados.

In [10]:
df["education"].value_counts()

education
secondary    25229
tertiary     13034
primary       6725
?              226
Name: count, dtype: int64

## 7. Nível de Escolaridade

A maioria dos clientes possui educação secundária (secondary), seguida por educação superior (tertiary) e educação primária (primary).

Foram identificados 226 registos com o valor "?", indicando informação em falta ou não normalizada.

A repetição deste valor em múltiplas variáveis sugere a existência de um conjunto consistente de registos com dados incompletos.

In [11]:
df.isna().sum()

unnamed: 0    0
age           0
job           0
marital       0
education     0
default       0
balance       0
housing       0
loan          0
contact       0
day           0
month         0
duration      0
campaign      0
pdays         0
previous      0
poutcome      0
y             0
dtype: int64

## 8. Valores em Falta

A análise dos valores nulos demonstra que não existem valores ausentes após o processo de limpeza e carregamento para PostgreSQL.

No entanto, foram identificados valores categóricos como "?" e valores numéricos potencialmente inválidos, indicando que a ausência de valores nulos não garante necessariamente a qualidade total dos dados.

Assim, a validação dos dados deve considerar não apenas valores nulos, mas também valores inconsistentes ou fora dos intervalos esperados.

# Conclusão

A análise exploratória dos dados permitiu compreender a estrutura e qualidade do dataset Bank Marketing após o processo de limpeza e ingestão para PostgreSQL.

O dataset contém 45.214 registos e 18 variáveis, não apresentando valores nulos após o tratamento efetuado.

A análise revelou que a maioria dos clientes é casada, possui educação secundária e pertence às categorias profissionais blue-collar, management e technician.

Foram também identificadas várias inconsistências nos dados, incluindo:

- Valores categóricos representados por "?"
- Diferentes representações para a variável alvo (`yes`, `no`, `subscribed`, `not-subscribed`)
- Valores numéricos potencialmente inválidos, como idades negativas ou excessivamente elevadas
- Presença da coluna `unnamed: 0`, provavelmente resultante da exportação do dataset original

Estas observações demonstram a importância da EDA na validação da qualidade dos dados, permitindo identificar situações que podem não ser detetadas apenas através dos processos automáticos de limpeza.